internal/generator/polynomial/template/polynomial.go.tmpl

```diff
@@ -200,3 +201,120 @@ func (p Polynomial) Text(base int) string {
 
 	return builder.String()
 }
+
+// InterpolateOnRange maps vector v to polynomial f
+// such that f(i) = v[i] for 0 ≤ i < len(v).
+// len(f) = len(v) and deg(f) ≤ len(v) - 1
+func InterpolateOnRange(v []{{.ElementType}}) Polynomial {
+	nEvals := uint8(len(v))
+	if int(nEvals) != len(v) {
+		panic("interpolation method too inefficient for nEvals > 255")
+	}
+	lagrange := getLagrangeBasis(nEvals)
+
+	var res Polynomial
+	res.Scale(&v[0], lagrange[0])
+
+	temp := make(Polynomial, nEvals)
+
+	for i := uint8(1); i < nEvals; i++ {
+		temp.Scale(&v[i], lagrange[i])
+		res.Add(res, temp)
+	}
+
+	return res
+}
+
+// lagrange bases used by InterpolateOnRange
+var (
+	lagrangeBasisLock sync.RWMutex
+	lagrangeBasis     [][]Polynomial
```

Comment:
I would just use a sync.Map here. Unless we expect a scenario were lots of go routine wants to read while others may initialize entries with different domainSize, a sync.Map Load / Store is clean enough

Reply:
Is it called in a "very concurrent way" ? because w/ current design ,the cache if you have 2 go routines querying domain size 100, not init yet; they will both hit the Rlock, and both compute the entry, and both update the lagrangeBasis array. 

so either mutex everything (i;e. the full getter is protected by a Lock / Unlock), or have one lock per entry if we expect concurrent accesses  on different domain sizes

Reply:
We expect very few writes (identical ones too, so imo "pre" computing size 100 twice is not a huge deal) and many, many reads that shouldn't block each other. I think a regular mutex wouldn't be good for the use case but sync.Map would be.